### PR TITLE
ISSUE-433: Behavior of rules sharing conditions within :or is incorrect

### PR DESCRIPTION
**Pull request by EthanEChristian**
For #433, i think this is the most straight forward approach. 

cc. @WilliamParker  @mrrodriguez

### CODE REVIEW

File: src/main/clojure/clara/rules/compiler.clj [hunk 1]
```diff
@@ -1138,18 +1140,21 @@
             ;; will out perform equivalence checks.
             update-node->id (fn [m id]
                               (let [node (get id-to-condition-nodes id)
-                                    prev (get m node)]
+                                    ;; appending parents to allow for identical conditions that do not share the same
+                                    ;; parents, see Issue 433 for more information
+                                    node-with-parents (assoc node :parents (get backward-edges id))
```
Comment (mrrodriguez) [inline src/main/clojure/clara/rules/compiler.clj]: I'm trying to rationalize this.

If we had a case like:
```
a   b
| |
c
| |
d e
```
where `c` was a common condition with, `a` and `b` as parents and `d` and `e` as children, it isn't safe to reuse the _same_ `c` to propagate to the children `d` and `e` in all cases. Instead `c` needs to propagate to `d`, eg. _only_ when it is receiving propagation from it's parent `a`, and similarly, perhaps `e` should only be propagated to when we are going from `b` > `c` > `e`.

I think this makes sense. `c` can't be shared in the propagation - or even working memory structure, due to propagation semantics - when it is coming from 2 logically separate parent paths.

So far I'm onboard with this change.

File: src/main/clojure/clara/rules/compiler.clj [hunk 2]
```diff
@@ -1140,16 +1140,27 @@
                               (let [node (get id-to-condition-nodes id)
                                     prev (get m node)]
                                 (if prev
-                                  (assoc! m node (min prev id))
-                                  (assoc! m node id))))
+                                  ;; There might be nodes with multiple representation under the same parent, this would
+                                  ;; likely be due to nodes that have the same condition but do not share all the same
+                                  ;; parents, see Issue 433 for more information
+                                  (assoc! m node (conj prev id))
+                                  (assoc! m node [id]))))
```
[the remark this answers is not in the document]
Comment (EthanEChristian) [reply src/main/clojure/clara/rules/compiler.clj]: these ids don't get cast to a set, rather we iterate over them and grab their parents for comparison.

[the remark this answers is not in the document]
Comment (mrrodriguez) [reply src/main/clojure/clara/rules/compiler.clj]: Good point, however, that made me realize that the `(set parent-ids)` should be done once in the let-binding, not every iteration of the `some` as it is now done below. 💀

File: src/main/clojure/clara/rules/compiler.clj [hunk 3]
```diff
@@ -1140,16 +1140,27 @@
...
+
+            backward-edges (:backward-edges beta-graph)
 
             ;; Use the existing id or create a new one.
-            node-id (or (get node->id node)
+            node-id (or (when-let [common-nodes (get node->ids node)]
+                          ;; We need to validate that the node we intend on sharing shares the same parents as the
+                          ;; current node we are creating. See Issue 433 for more information
+                          (some #(when (= (get backward-edges %)
+                                          (set parent-ids))
```
Comment (mrrodriguez) [inline src/main/clojure/clara/rules/compiler.clj]: See comments under https://github.com/cerner/clara-rules/pull/434/files#r323788944

File: src/test/common/clara/test_simple_rules.cljc [hunk 4]
```diff
@@ -194,3 +194,31 @@
     (is (has-fact? (:cold @side-effect-holder) (->Temperature -10 "MCI") ))
 
     (is (has-fact? (:subzero @side-effect-holder) (->Temperature -10 "MCI")))))
+
+
```
[the remark this answers is not in the document]
Comment (EthanEChristian) [reply src/test/common/clara/test_simple_rules.cljc]: yes, the e-query assertion fails prior to the changes in the PR

**Pull request merged**